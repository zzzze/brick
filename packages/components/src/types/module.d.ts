// eslint-disable-next-line
/// <reference path="../../../../node_modules/monaco-editor/monaco.d.ts" />

declare const __DEV__: boolean

declare namespace NodeJS {
  interface Global {
    __DEV__: boolean
  }
}

declare module '*.md' {
  const Component: React.ReactElement
  export default Component
}
