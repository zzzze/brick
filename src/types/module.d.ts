declare module '*.md' {
  const Component: React.ReactElement
  export default Component
}
