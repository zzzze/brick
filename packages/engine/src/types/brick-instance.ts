export interface SetDataFn {
  (data: DataObject): DataObject
}

export interface SetDataOptions {
  setToBlueprint?: boolean // whether set data into blueprint, if false set data at runtime only
}

export interface SetData {
  (fn: SetDataFn, options: SetDataOptions): void
}

export interface BrickContext {
  data?: Record<string, unknown>
  actions?: Record<string, unknown>
}

export interface Action {
  (...args: unknown[]): void
  __source?: string
  __originAction?: Action
  binded?: boolean
  instance_key?: string
}

export type DataObject = Record<string, unknown>

export interface Emit {
  (event: string, ...args: unknown[]): void
}

export interface BrickInstance {
  key: string
  data: DataObject
  actions: Record<string, Action>
  handlers: Record<string, Action>
  listeners: Record<string, Action>
  context: BrickContext
  setData: SetData
  emit: Emit
  editing?: boolean
  children?: React.ReactNode
}
