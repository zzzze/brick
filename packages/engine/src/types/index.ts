export * from './brick'
export * from './config'
export * from './engine'
export * from './brick-instance'
