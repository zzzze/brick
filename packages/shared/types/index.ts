export * from './form'

export type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>
}
