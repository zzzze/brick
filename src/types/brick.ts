export enum PropType {
  STRING = 'string',
  NUMBER = 'number',
  BOOLEAN = 'boolean',
}

export enum ChildrenType {
  SINGLE = 'single',
  MULTIPLE = 'multiple',
  NONE = 'none',
}

interface RenderProps {
  value: PropsObject
  children?: JSX.Element | JSX.Element[]
}

interface ConfigFormRenderProps {
  value: PropsObject
  onChange: (value: PropsObject) => void
  configFormVisible?: boolean
  hideConfigForm?: () => void
  children?: JSX.Element | JSX.Element[]
}

export type PropsObject = Record<string, unknown>

export interface Brick {
  name: string
  icon: string
  propTypes: Record<string, PropType>
  defaultProps: PropsObject
  childrenType: ChildrenType
  renderConfigForm: React.FC<ConfigFormRenderProps> | React.ComponentType<ConfigFormRenderProps>
  render: React.FC<RenderProps> | React.ComponentType<RenderProps>
  version: string
}
