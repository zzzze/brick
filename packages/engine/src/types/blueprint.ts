import { Func } from '@/action/compile-action'
import { Render } from './brick'
import { DataObject } from './brick-instance'

export const idPrefix = '$'

export interface Supply {
  data?: Record<string, unknown>
  actions?: Record<string, string>
}

export interface CustomRenderFn {
  (): Render
}

export type CustomRender = string | CustomRenderFn

export interface Blueprint {
  _key: string
  name: string // brick name
  version: string // brick version
  id?: string // brick instance id
  data?: DataObject // data for brick instance
  actions?: Record<string, Func> // define action which can be used by supply or handler
  handlers?: Record<string, Func> // define handler of event that triggered by brick instance
  supply?: Supply // provide data for child brick instance
  children?: Blueprint[]
  listeners?: Record<string, Func> // listeners use for register event listeners of EventEmitter
  render?: CustomRender // custom render
}

export interface SetBlueprintFn {
  (config: Readonly<Blueprint>): Blueprint
}

export interface SetBlueprint {
  (fn: SetBlueprintFn, cb?: () => void): void
}
