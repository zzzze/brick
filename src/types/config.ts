import { PropsObject } from './brick'

export interface Config {
  name: string
  props: PropsObject
  children?: Config[]
  version: string
}
