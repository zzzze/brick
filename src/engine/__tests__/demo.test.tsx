import React from 'react'
import demoTest from '@/tools/tests/demoTest'
import Engine from '@/engine'
import BrickContainer from '@/engine/brick-containter'
import { Brick, ChildrenType, ConfigFormRenderProps, PropType, RenderProps } from '@/types'

const View: Brick = {
  name: 'View',
  propTypes: {},
  defaultProps: {},
  childrenType: ChildrenType.MULTIPLE,
  renderConfigForm(props: ConfigFormRenderProps) {
    return <BrickContainer>edit View: {props.children}</BrickContainer>
  },
  render(props: RenderProps) {
    return <BrickContainer>{props.children}</BrickContainer>
  },
  version: '0.0.1',
}

const Text: Brick = {
  name: 'Text',
  propTypes: {
    content: PropType.STRING,
  },
  defaultProps: {
    content: '',
  },
  childrenType: ChildrenType.NONE,
  renderConfigForm(props: ConfigFormRenderProps) {
    return <BrickContainer tag="span">edit Text: {props.value.content as string}</BrickContainer>
  },
  render(props: RenderProps) {
    return <BrickContainer tag="span">{props.value.content as string}</BrickContainer>
  },
  version: '0.0.1',
}

Engine.registerBrick(View)
Engine.registerBrick(Text)

demoTest('engine')
