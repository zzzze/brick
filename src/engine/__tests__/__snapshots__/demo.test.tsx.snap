// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ./src/engine/demo/basic.md correctly 1`] = `
<span>
  <div>
    <button
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  hello world
</span>
`;

exports[`renders ./src/engine/demo/children.md correctly 1`] = `
<div>
  <div>
    <button
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <span>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    hello
  </span>
  <span>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    world
  </span>
</div>
`;

exports[`renders ./src/engine/demo/config-tree-array.md correctly 1`] = `
Array [
  <span>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    hello
  </span>,
  <span>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    world
  </span>,
  <div>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span>
      <div>
        <button
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      hello
    </span>
    <span>
      <div>
        <button
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      world
    </span>
  </div>,
]
`;

exports[`renders ./src/engine/demo/custom-render.md correctly 1`] = `
<div>
  <div>
    <button
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <span>
    foo
  </span>
  <span>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    bar
  </span>
</div>
`;

exports[`renders ./src/engine/demo/supply.md correctly 1`] = `
<div>
  <div>
    <button
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <div>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span>
      <div>
        <button
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      foo
    </span>
  </div>
</div>
`;

exports[`renders ./src/engine/demo/supply-from-data.md correctly 1`] = `
<div>
  <div>
    <button
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <div>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span>
      <div>
        <button
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      bar
    </span>
  </div>
</div>
`;

exports[`renders ./src/engine/demo/supply-with-id.md correctly 1`] = `
<div>
  <div>
    <button
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <div>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span>
      <div>
        <button
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      foo
    </span>
  </div>
</div>
`;

exports[`renders ./src/engine/demo/supply-without-id.md correctly 1`] = `
<div>
  <div>
    <button
      data-testid="edit-btn"
    >
      edit
    </button>
  </div>
  <div>
    <div>
      <button
        data-testid="edit-btn"
      >
        edit
      </button>
    </div>
    <span>
      <div>
        <button
          data-testid="edit-btn"
        >
          edit
        </button>
      </div>
      baz
    </span>
  </div>
</div>
`;
